### ISSUE TITLE
MLLM ​for Generating Industrial Computer Vision Workflows​

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#25]
- We'll put more specific project content here

**Comment by xp1632** [comment]
### Poster Comments

poster content: https://drive.google.com/drive/folders/1u6r0foca6H_gA9RIsSWWPsjtUXMWfnCA?usp=drive_link

----

- **Content Side**:
1. Domain experts are not our only audience
2. Our customers might not be lacking programming knowledge, so no expertise gap

----

- **General Side**:
1. Use more passive tense.
2. Substitute `Traditional` with `Conventional` when describing technology

----

- **Things To Notice**:
1. Our selling point could be `cost` and `energy` side
2. The generated workflow is `classical`, `filter-based`, so don't over exaggerate it

**Comment by xp1632** [comment]
### Vision Language Models 

- we choose BLIP-2 now over CLIP [github](https://github.com/openai/CLIP/blob/main/model-card.md), [model card](https://huggingface.co/docs/transformers/en/model_doc/clip)
- and we'll check LLava afterwards [github](https://llava-vl.github.io/), [model card](https://huggingface.co/docs/transformers/en/model_doc/llava)


---

### BLIP-2 

- [github](https://github.com/salesforce/LAVIS/tree/main/projects/blip2)
- [paper](https://arxiv.org/pdf/2301.12597)
- [model card](https://huggingface.co/docs/transformers/en/model_doc/blip-2)
- [x] Run with local GPU

- we've successfully run `blip2-flan-t5-xl`
- with Pytorch version: 2.5.1, CUDA 12.1, NVIDIA GeForce RTX 4070 Ti SUPER

![Image](https://github.com/user-attachments/assets/44031c32-6820-4fcb-aa11-4f7f0d04b884)

**Comment by xp1632** [comment]
### Dataset


- **We chose the first crack dataset**
- As we can see, there are no description or _certain types of the crack_
- we first try simple text descriprtions such as: `one horizontal crack` , `one vertical crack`, then we go to more complex ones such as `veins of leaf type of crack`


----

- We need crack image dataset， we'll choose from the following:

1.  11.200 images that are merged from 12 available crack segmentation datasets.
https://www.kaggle.com/datasets/lakshaymiddha/crack-segmentation-dataset

![Image](https://github.com/user-attachments/assets/04eeea13-d6aa-46c3-937a-5ac3ddb151fb)

2. or Crack-Detection-and-Segmentation-Dataset-for-UAV-Inspections
https://github.com/KangchengLiu/Crack-Detection-and-Segmentation-Dataset-for-UAV-Inspection
![Image](https://github.com/user-attachments/assets/e691471c-6451-4418-8991-912ad6e2a222)

3. or Concrete Crack Images for Classification (only positive or negative)
https://www.kaggle.com/datasets/arnavr10880/concrete-crack-images-for-classification/data?select=Positive

![Image](https://github.com/user-attachments/assets/00d9f1a3-8600-4b65-9541-5f3a43823306)

**Comment by xp1632** [comment]
### Code Generation Model

![Image](https://github.com/user-attachments/assets/fe9e9bcf-adb4-4c85-b88e-23bd8defe7cb)


- We have several options for generating Python code, APIs or local model
- after comparison, we choose `Code Llama-Python` as first attempts [github](https://github.com/meta-llama/codellama)

![Image](https://github.com/user-attachments/assets/89aea4f2-c005-4c6b-ad8a-dd67a845362e)
[image refer](https://ai.meta.com/blog/code-llama-large-language-model-coding/)

—-


### Code Llama

- EU restrictions: 
    - when we check the GitHub page and paper for Lllama model, it doesn’t mention there are restrictions for the usage especially for EU users

- for legal reasons, we need to find an alternative open source code generation model
- we still file in an official request just to see if it’s a blurred regulation
- the access is granted by the developer when stating as an EU user, whatever


**Ollama**

- or we can access code llama and other mlllms via Ollama

**Comment by xp1632** [comment]
### Workflow Composition


### **- Theory part:** 
- [x]  1. **What are the roles for Blip2 and Llama?** 


![Image](https://github.com/user-attachments/assets/cffa4682-1235-4319-8f87-7ce3b12db35a)

- Blip2 infers the visual feature and combines it with text
- Llama uses the more concrete text description generated by Blip2 and generate more specific code

---

- [ ] 2. **Is this combination of Vision Language Model and Code Generation Model new?**

- No, we can find many works with this aim (intersection of multimodal understanding and code generation.):
-  GPT-4V (OpenAI, 2023) [system card](https://openai.com/index/gpt-4v-system-card/)
-  CodeGen-Vision (Zhang et al., 2023)


---
### **- Technical Part:**

**0. How should we let Blip2 model process the three inputs: `sample image`, `text requirements`, `image dataset`?** 



--- 

**1. How do we calculate the similarity between vision/text joint features we get from Blip2 model?**


---


**2. What items should we include in the enhanced information/prompt JSON file that we generated from Blip2 model and will feed Llama model?**


---


**3. Why do we get wrong characteristic answers from model `blip2-flan-t5-xl`?** 

![Image](https://github.com/user-attachments/assets/69a2a366-7717-4498-b866-5d6e2a8dc058)

![Image](https://github.com/user-attachments/assets/3bfa2829-184b-4294-80c2-b6e80ba0ab84)

>  is it a RGB image?   No


- [x] - (x)3.1 is it because we shouldn't ask like this?
- We rephrased the question and still get inconsistent answers

![Image](https://github.com/user-attachments/assets/69a39b60-aab9-42a0-bbc6-3201bfe4d195)



- [x] - (x) 3.2 did we lose some color channel when processing the image before inputting it to model?
- the processed image keeps shape of color channel of RGB image as [1,3,height, width], so no color channel is lost

![Image](https://github.com/user-attachments/assets/1bb178ce-6be2-45c1-a78b-dde762711054)

- we also visualized the image, nothing is lost
![Image](https://github.com/user-attachments/assets/e21d8337-5f93-45bf-8f6f-09098bddc81d)

- [x] - (v)3.3 Is it because blip2 model's vision model part ViT, is trained to focus on extracting high-level semantic features rather than preserving low-level image statistics?

- so when fine-tuned on datasets like COCO, these models learn to identify objects, scenes, and their relationships, but not necessarily to answer technical questions about image properties.


- [x] -  (-) 3.4Is it because we are using the flan version model and need more accuracy?






- [x] - (v) 3.5 How should we get a more accurate answer about the meta characteristic information from the image?


   - Method 1: Get meta information with OpenCV/PIL with pre-analysis
   

     > image = Image.open(requests.get(image_url, stream=True).raw)
     > is_color = image.mode == "RGB"

  - Method 2: fine-tuning the model on custom dataset with examples of RGB and grayscale images and explicit questions/correct answers about image properties

**Comment by xp1632** [comment]
### Cluster 


- [x] 1. Registration:
-  https://wiki.bwhpc.de/e/DACHS/Registration
- We've finished the registration to use bwUniCluster and DACHS 



---

- [x] 2. Slurm 


- [x] SSH login with MobaXterm when using Windows system
- https://wiki.bwhpc.de/e/DACHS/Login

- 2.1 MobaXterm settings:
     - create new SSH Session in MobaXterm
     - <username> should be prefixed with local username; for Aalen University, the prefix is `aa`, so it should be `aa_06xxx`
![Image](https://github.com/user-attachments/assets/a50f723c-2997-4fe9-a488-64eaad22caea)

- 2.2 `OTP` One Time Password 

    - it requires OTP for logging in the server, we need to register one somehow somewhere
    - when trying to register OTP for cluster, the link itself requires a OTP code which the schools didn't work
    - have contacted Wolfgang for advice
    - Wolfgang said he'll send back info when he figures out. (He sent me an invitation for DACHS seminar Mar. 14th, 14:00)
    - I think my another option is to do the `Lost Token` part
    https://wiki.bwhpc.de/e/Registration/2FA#Lost_Token 
    - file a ticket https://bw-support.scc.kit.edu/#ticket/view/my_tickets


- [x] The ticket of requesting resetting unknown Token is processed efficiently, they disabled the wrong token and I was able to add new correct ones

![Image](https://github.com/user-attachments/assets/ea10ebe7-1368-4dba-9370-59c3acbf5bf4)

- 2.3 add SSH key




---


- [x] 3. Running Calcalation:
- logging in with OTP and service password we set in [the ](https://login.bwidm.de/)
![Image](https://github.com/user-attachments/assets/ee9a7812-1dda-495e-885e-47b5a7ba5322)

- https://wiki.bwhpc.de/e/Running_Calculations

-**ollama for llm model**
- https://ollama.com/library
https://wiki.bwhpc.de/e/Development/ollama

---

- [ ] 4. Workshop of DACHS:

- whole workshop of bash, slurm, workspace, ollama: https://www.hs-esslingen.de/informatik-und-informationstechnik/forschung-labore/forschung/laufende-projekte/dachs 

- the pdf slides and certain part videos of using this cluster are released:

Die Folien sind als PDF verfügbar auf:
- https://www.hs-esslingen.de/fileadmin/media/Fakultaeten/it/FORSCHUNG_LABORE/DACHS/DACHS-Workshop-2025.03.14.pdf


Auch werden die Videos dort verlinkt. Bis dahin können Sie die Videos anschauen unter:
  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_1_Introduction_Overview.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_2_Entitlement_Registration.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_3_OTP_SSH_Keys.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_4_Linux_Command_Line.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_5_Slurm.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_6_Workspace_Tools.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_7_Ollama.mov

  https://www2.hs-esslingen.de/~rakeller/2025.03.14-DACHS_Workshop_8_More_Info.mov

**Comment by xp1632** [comment]
### Thoughts about whole workflow


- While we are doing the branch project, we don't want to lose the thoughts and tracks for the whole workflow, here's the track that we can pick up later: 


1. Whole project design 
- detailed:  Multimodal Crack Image Retrieval: https://claude.ai/share/d6dd5057-1e02-4960-8dc3-d212f7e80df5
- modularized : Multimodal Workflow for Leaf Vein Crack Detection: https://claude.ai/share/f56d8214-ff57-4f99-8911-b6fd8025d5b9



2. what information blip provide from vision feature: 

- Identifying Leaf Crack Patterns with BLIP-2 : https://claude.ai/share/dc0133bd-d3f8-44a0-83bc-b58dbd409b8b

3. blip2 insight --> generating code - Generating Python Image Processing Code from BLIP2 Insights: https://claude.ai/share/5a7ce485-bd2d-436d-af34-4a2da82092d5
